### PR TITLE
CloudWatch task handler doesn't fall back to local logs when Amazon CloudWatch logs aren't found

### LINKED ISSUE #27396: CloudWatch task handler doesn't fall back to local logs when Amazon CloudWatch logs aren't found
This is really a CloudWatch handler issue - not "airflow" core.

### Discussed in https://github.com/apache/airflow/discussions/27395

<div type='discussions-op-text'>

<sup>Originally posted by **matthewblock** October 24, 2022</sup>
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

We recently activated AWS Cloudwatch logs. We were hoping the logs server would gracefully handle task logs that previously existed but were not written to Cloudwatch, but when fetching the remote logs failed (expected), the logs server didn't fall back to local logs.

```
*** Reading remote log from Cloudwatch log_group: <our log group> log_stream: <our log stream>
```

### What you think should happen instead

According to documentation [Logging for Tasks](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/logging-tasks.html#writing-logs-locally), when fetching remote logs fails, the logs server should fall back to looking for local logs:

> In the Airflow UI, remote logs take precedence over local logs when remote logging is enabled. If remote logs can not be found or accessed, local logs will be displayed.

This should be indicated by the message `*** Falling back to local log`.

If this is not the intended behavior, the documentation should be modified to reflect the intended behavior.

### How to reproduce

1. Run a test DAG without [AWS CloudWatch logging configured](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/logging/cloud-watch-task-handlers.html)
2. Configure AWS CloudWatch remote logging and re-run a test DAG

### Operating System

Debian buster-slim

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)
</div>

### PR DESCRIPTION
**Pull request by hankehly**
Closes #27396

### Summary

The CloudWatch task log handler does not default to local logs (like the documentation says it should) when remote logs are unavailable. This PR fixes the issue.

### Todo
- [x] Update CloudWatch log handler
- [x] Add/update unit tests
- [x] Manual check (see below for details)

#### Procedure

1. Run a test DAG with [AWS CloudWatch logging configured](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/logging/cloud-watch-task-handlers.html)
2. Verify that logs are sent to CloudWatch
3. Delete CloudWatch log stream associated with task
4. Create log file locally `echo "test" >> {log file path}`
5. Refresh the task log page

#### Expectations

* "Unable to read remote logs from Cloudwatch" should appear in the log because we deleted the log stream associated with the task.
* "Reading local file: ..." should appear in the log to show that the log handler falls back to local logs.
* The contents of the local log file should appear (in this case the contents are "test")

#### Screenshot

Screenshot 2022-11-09 at 7 09 49

### CODE REVIEW

**Review by vincbeck — APPROVED** [review]
Nice!

**Comment by hankehly** [comment]
@vincbeck 
Thanks for the early review. Once I've handled the failing tests I'll click the "Ready for review" button.

File: tests/providers/amazon/aws/log/test_cloudwatch_task_handler.py [hunk 1]
```diff
@@ -154,49 +154,22 @@ def test_read(self):
             [{"end_of_log": True}],
         )
 
-    def test_read_wrong_log_stream(self):
```
Comment (hankehly) [inline tests/providers/amazon/aws/log/test_cloudwatch_task_handler.py:157]: I deleted these 2 test cases for the following reasons.
* They overlap with the one I added (testing for generic failure to fetch remote logs)
* AWS error messages are now propagated to the task log. These strings could change in the future and break our unit tests.

File: airflow/providers/amazon/aws/log/cloudwatch_task_handler.py [hunk 2]
```diff
@@ -97,17 +107,12 @@ def get_cloudwatch_logs(self, stream_name: str) -> str:
...
         :return: string of all logs from the given log stream
         """
-        try:
-            events = self.hook.get_log_events(
-                log_group=self.log_group,
-                log_stream_name=stream_name,
-                start_from_head=True,
-            )
-            return "\n".join(self._event_to_str(event) for event in events)
-        except Exception:
```
Comment (hankehly) [inline airflow/providers/amazon/aws/log/cloudwatch_task_handler.py:107]: I am propagating the AWS error message to the task log. I think this will be useful for debugging IAM/connection issues.

File: tests/providers/amazon/aws/log/test_cloudwatch_task_handler.py [hunk 3]
```diff
@@ -154,49 +154,22 @@ def test_read(self):
...
+            mock_get_logs.side_effect = Exception("Failed to connect")
+            log, metadata = self.cloudwatch_task_handler._read(self.ti, self.ti.try_number)
+        expected_log = (
+            f"*** Unable to read remote logs from Cloudwatch (log_group: {self.remote_log_group}, "
+            f"log_stream: {self.remote_log_stream})\n*** Failed to connect\n\n"
+            # The value of "log_pos" is equal to the length of this next line
+            f"*** Reading local file: {self.local_log_location}/{self.remote_log_stream}\n"
         )
+        assert log == expected_log
+        expected_log_pos = 26 + len(self.local_log_location) + len(self.remote_log_stream)
```
Comment (hankehly) [inline tests/providers/amazon/aws/log/test_cloudwatch_task_handler.py:170]: Log position seems irrelevant in the context of the CloudWatch handler because we fetch/return all logs in one request. I'm following the example I found in the GCS task handler here.

https://github.com/apache/airflow/blob/21063267fd9764b2ca38669e8faec75d9b87179c/tests/providers/google/cloud/log/test_gcs_task_handler.py#L108

**Review by potiuk — APPROVED** [review]

**Comment by o-nikolas** [comment]
:rocket:
